src/ddsrt/src/xmlparser.c

```diff
@@ -192,7 +192,9 @@ static int make_chars_available (struct ddsrt_xmlp_state *st, size_t nmin)
     /* ensure buffer space is available */
     if (st->fp != NULL) {
         if (pos + nmin > st->cbufmax) {
-            memmove (st->cbuf, st->cbuf + pos, st->cbufn - pos);
```

Comment:
I have difficulty accepting this change, but I can live with it on the grounds that the code doesn't come with a *proof* that `0 <= pos <= st->cbufn`.

Still, I would prefer at least a comment explaining why this test is here (so a reference to MISRA 21.18).

However, that is not the last of it. If `pos == cbufn`, then fine. If `pos > cbufn`, which must be assumed possible if one requires this test because there is no proof that `pos <= cbufn`, then the values of `cbufn` and `cbufp` will become huge (on a 2 complement machine), and certainly *much* larger than the underlying buffer, and so skipping the `memcpy` to avoid undefined behaviour means we then have a problem on *all* memory accesses of `st->cbuf`. Checking those is infeasible in C. If only this were Ada and not C ...

So the only sensible action on encountering `st->cbufn < pos` is to call `abort()`. But, in all there wisdom, the MISRA authors decided that aborting execution is not allowed, not even when you know there is no recovery possible ...